### PR TITLE
Token parsing

### PR DESCRIPTION
**Pull request by erikrose**
Here's my simplified take on token parsing, based on @bloff's work on #40 (which this supercedes but which has interesting discussion on it that you should read). Thoughts?

### CODE REVIEW

File: parsimonious/grammar.py [hunk 1]
```diff
@@ -434,6 +442,21 @@ def visit_rules(self, node, (_, rules)):
...
+    pre-lexed tokens rather than strings"""
+
+    def visit_spaceless_literal(self, spaceless_literal, visited_children):
+        """Turn a string literal into a ``TokenMatcher`` that matches
+        ``Token`` objects by their ``type`` attributes."""
+        return TokenMatcher(evaluate_string(spaceless_literal.text))
+
+    def visit_regex(self, regex, (tilde, literal, flags, _)):
+        raise BadGrammar('Regexes do not make sense in TokenGrammars, since '
+                         'they operate on pre-lexed tokens rather than '
```
Comment (erikrose) [inline parsimonious/grammar.py]: "they" is ambiguous (at best) here.

File: parsimonious/nodes.py [hunk 2]
```diff
@@ -34,7 +34,7 @@ class Node(StrAndRepr):
     __slots__ = ['expr_name',  # The name of the expression that generated me
                  'full_text',  # The full text fed to the parser
                  'start', # The position in the text where that expr started matching
-                 'end',   # The position after start where the expr first didn't
+                 'end',   # The position after starft where the expr first didn't
```
Comment (keleshev) [inline parsimonious/nodes.py]: lol

**Comment by keleshev** [comment]
I like this a lot, especially the fact that you can write `"+"` instead of `PLUS` or `%PLUS%`.

One concern: imagine you are trying to parse a programming language with reserved keywords. PEGs usually alleviate the need for reserved keywords but, say, you want to conform to existing language spec.
In token-less PEG you do something like this:

```peg
reserved_keyword -> "def" | "class"
identifier -> !reserved_keyword [a-zA-Z_]+
```

How would you represent this with tokenful PEG?

**Comment by erikrose** [comment]
```
reserved_keyword = "def" / "class"
identifier = !reserved_keyword "name"
```

`name` is a "name"-type token. It would have been tokenized by whatever pre-processing phase we're talking about, prior to calling Parsimonious. It would probably have a `text` attr or something containing its actual char string.

In other words, I'm proposing an entirely tokenized input sequence, not some hybrid of tokens and string chunks.

**Comment by keleshev** [comment]
> It would probably have a text attr or something containing its actual char string.

Got it now.

**Comment by erikrose** [comment]
I'm not completely sold on this, but I'm going to merge it in as an experimental feature. I am persuaded that *some* kind of token-based parsing is useful, but I'm not sure what kind until we have a couple of concrete use cases. I'm working on an indentation-significant language right now that might serve as an educational test case, so we'll see what comes out of that.

**Pull request merged**